Fix fatal error on sites using WooCommerce Subscription versions below 4.0.0

malithsen opened the pull request:
#### Changes proposed in this Pull Request

Changes introduced in #6344 relies on `WCS_Staging::is_duplicate_site()`. This function was introduced in WCS 4.0.0. Sites using older versions of WCS are encountering a fatal error.

This PR attempts to fix it by using `WC_Subscriptions::is_duplicate_site` for pre 4.0.0 WCS versions. While this prevents the fatal error, it is unclear to me whether this is the correct usage.



<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Install and activate WooCommerce Subscriptions 3.1.5 (or any version below 4.0.0)
* Make sure the site is still up and the following fatal error is not thrown

```
Uncaught Error: Call to undefined method WCS_Staging::is_duplicate_site() in /srv/htdocs/wp-content/plugins/woocommerce-payments/includes/subscriptions/class-wc-payments-subscriptions.php:132
```

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

malithsen committed:
Use the correct class for WCS versions below 4.0.0

malithsen committed:
Remove unnecessary check

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `6489` or branch name `fix/pre-wcs-4.0.0-fatal-error` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=fix/pre-wcs-4.0.0-fatal-error) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 848433366e9ec02351384dea2ba13b5e86bcf4f4
- Build time: 2023-06-28 16:20:40 UTC

_Note: the build is updated when a new commit is pushed to this PR._

mattallan reviewed (commented):
Thanks @malithsen for making these changes.
The code changes look good and I've tested the following cases:

- [x] Installed WC Subscriptions 3.1.6, confirmed fatal on `develop` and fixed with these changes
- [x] With WC Subscriptions 3.1.6, I confirmed the WCPay subscription is not updated in Stripe when site is in staging mode and confirmed it is when staging is off
- [x] With latest WC Subscriptions (same tests above i.e. staging + not staging)
- [x] Without WC Subscriptions (i.e. just WCS Core bundled) confirmed the same tests.

james-allan reviewed (approved):
This looks good to me @malithsen. 💯 

I tested a following situations:

1. **Pre WC Subscriptions 4.0 (v 3.1.6).** 
   - Staging enabled 🛑 - subscription events **don't** get sent to Stripe. 
   - Not staging 🟢 - subscription events **do** get sent to Stripe. 
   - This `is_duplicate_site()` change is using `WC_Subscriptions::is_duplicate_site()` successfully.
2. **Post WC Subscriptions 4.0 (latest).** 
   - Staging enabled 🛑 - subscription events **don't** get sent to Stripe. 
   - Not staging 🟢 - subscription events **do** get sent to Stripe. 
   - This `is_duplicate_site()` change is still using `WCS_Staging::is_duplicate_site()` successfully on appropriate versions.

james-allan commented:
I also just ran a test by disabling the WC Subscriptions extension and like Matt mentioned in his review, it falls back to the `WCS_Staging` class packaged in WC Payments - subscriptions-core.

malithsen committed:
Fix unit tests

- Bump mocked WC_Subscriptions version so that both branches of
  is_duplicate_site are being tested
- mock is_duplicate_site method

malithsen committed:
Add changelog entry

malithsen committed:
Remove redundant variable

malithsen commented:
Thanks for the reviews @james-allan @mattallan 
I made a couple of small changes to the tests to get them passing. Can you please give another look at the code changes? This is not urgent, as this PR will now be going out in the next regular scheduled release.

james-allan commented:
Sorry @malithsen I missed the original ping. Those recent changes look good to me.

james-allan commented on changelog/fix-pre-wcs-4.0.0-fatal-error:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: fix
+
+Fix a fatal error on sites using using WCS versions below 4.0.0
```
```suggestion
Fix a fatal error on sites using WC Subscriptions versions below 4.0.0
```

Fixed a typo and clarified wcs is Subscriptions

malithsen committed:
Fixed a typo and clarified wcs is Subscriptions

Co-authored-by: James Allan <james.allan@automattic.com>

malithsen committed:
Merge branch 'develop' into fix/pre-wcs-4.0.0-fatal-error

malithsen commented:
@james-allan no worries. I updated the changelog. Merging
